webhook/resources/backup/mutator.go

```diff
@@ -65,28 +64,29 @@ func (b *backupMutator) Create(request *admission.Request, newObj runtime.Object
 		metaLabels = map[string]string{}
 	}
 
-	//check snapshot name, if not exist, return error
-
-	volumeName, isExist := backup.Labels[types.LonghornLabelBackupVolume]
-	if !isExist {
-		if backup.Spec.SnapshotName == "" {
-			err := errors.Wrapf(err, "cannot find the backup volume label for backup %v", backup.Name)
-			return nil, werror.NewInvalidError(err.Error(), "")
-		}
-
+	var volume *longhorn.Volume
+	if backup.Spec.SnapshotName != "" {
 		// Try to get volume name from snapshot
 		snapshot, err := b.ds.GetSnapshotRO(backup.Spec.SnapshotName)
 		if err != nil {
 			return nil, errors.Wrapf(err, "failed to get the snapshot %v", backup.Spec.SnapshotName)
 		}
 
-		volume, err := b.ds.GetVolumeRO(snapshot.Spec.Volume)
+		volume, err = b.ds.GetVolumeRO(snapshot.Spec.Volume)
 		if err != nil {
 			return nil, errors.Wrapf(err, "failed to get the volume %v of the snapshot %v of the backup %v", snapshot.Spec.Volume, snapshot.Name, backup.Name)
 		}
+	}
 
-		volumeName = volume.Name
-		metaLabels[types.LonghornLabelBackupVolume] = volumeName
+	//check snapshot name, if not exist, return error
+
+	volumeName, isExist := backup.Labels[types.LonghornLabelBackupVolume]
+	if !isExist {
+		if volume == nil {
+			err := errors.Wrapf(err, "cannot find the backup volume label for backup %v", backup.Name)
+			return nil, werror.NewInvalidError(err.Error(), "")
+		}
+		metaLabels[types.LonghornLabelBackupVolume] = volume.Name
 	}
```

Comment:
Why do we need to refactor this part?
After the change, it always tries to get Volume and Snapshot although `LonghornLabelBackupVolume` label is already existing and introduces necessary GET operations.

Reply:
Reorganized the logic and using a reusable volume getter.